[functions] Allow functions to pass runtime specific options

addisonj opened the pull request:
### Motivation

This commit adds a new argument for functions, customRuntimeOptions, which is passed to
funcions (as well as sources/sinks) that enables the ability to
customize the runtime.

This is added primarily to support the `KubernetesManifestCustomizer`
interface. This interface is intended, as the name indicates, allows for
customizing how the kubernetes manifests are generated before they
are sent off to the k8s cluster. 

This interface has a default implementation, which allows for
changing the namespace, labels, nodeSelector labels, and toleratations per function.

This interface is also pluggable, allowing for more customized
implementations. For example, the functions for a given tenant could be
mapped to different pools of compute for isolation.

### Modifications

For the CLI and protobufs, the modifications just involve plumbing through the new option, `customRuntimeOptions` through the relevant code. 

For kubernetes runtime, the modifications are fairly straight forward, adding a new configuration option, `kubernetesManifestCustomizerClassName` and `kubernetesManifestCustomizerConfig` which are options under the kubernetes runtime.

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

This commit adds some tests:

  - Adds tests for the kubernetes runtime with the default and custom implementation that validates the manifest files generated
- passes existing unit tests 

### Does this pull request potentially affect one of the following parts:

*If `yes` was chosen, please highlight the changes*

  - Dependencies (does it add or upgrade a dependency): no
  - The public API: no
  - The schema: no
  - The default values of configurations:  no
  - The wire protocol: yes, only of functions protobufs, adds a new field, customRuntimeOptions, which is a simple string
  - The rest endpoints: no
  - The admin cli options: yes, adds `custom-runtime-options` flags to the create/update commands for functions, sources, and sinks, which populates the above protobuf field
  - Anything that affects deployment: (yes / no / don't know)

### Documentation

  - Does this pull request introduce a new feature?  yes
  - If yes, how is the feature documented? yes, docs
  - If a feature is not applicable for documentation, explain why? 
  - If a feature is not documented yet in this PR, please create a followup issue for adding the documentation

addisonj commented:
This conflicts with #5398, will need rebased if that is merged first

jerrypeng commented on pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/KubernetesRuntimeFactory.java:
```diff
@@ -340,8 +353,21 @@ void validateMinResourcesRequired(Function.FunctionDetails functionDetails) {
+        if (!isEmpty(manifestCustomizerClassName)) {
+            manifestCustomizer = (KubernetesManifestCustomizer) Reflections.createInstance(manifestCustomizerClassName, ClassLoader.getSystemClassLoader());
+        } else {
+            manifestCustomizer = new DefaultKubernetesManifestCustomizer();
```
I would suggest that the default to be a no-op implementation.  Allowing users the change the certain kubernetes settings when submitting a function can be a security risk.

addisonj commented:
okay @jerrypeng I changed that, will see if I can get the tests happy, let me know if you have any other feedback!

addisonj commented:
I started in on re-working this... the changes introduced in #5404 broke one of the assumptions this made, which was that we create new instances of the functionAuthProvider, with a specific one for any overridden namespace. @jerrypeng do you see any problem with pushing the creation of the FunctionAuthProvider (via reflection) into the KubernetesRuntimeFactory? Or should we move the namespace used elsewhere? This provides a lot more flexibility to end users as being able to isolate various tenants to their own k8s namespace for functions can be use for finer grain isolation of underlying resources.

Would love to get this one through this week, but just want to make sure I am aligned in how to go about this

addisonj commented:
rerun java8 tests

addisonj commented:
rerun integration tests

addisonj commented:
@jerrypeng @sijie I reworked this to account for the changes from #5404, I think it seems pretty reasonable.

If you could give another review and let me know if there is anything else you want to see here.

sijie commented:
@addisonj LGTM. Great change!

btw can you rebase this pull request?

addisonj commented:
rerun java8 tests

addisonj commented:
@sijie @jerrypeng this is ready for another look, let me know if you have any questions!

jerrypeng commented on pulsar-functions/worker/src/main/java/org/apache/pulsar/functions/worker/FunctionRuntimeManager.java:
```diff
@@ -151,7 +151,7 @@ public FunctionRuntimeManager(WorkerConfig workerConfig, WorkerService workerSer
         }
         secretsProviderConfigurator.init(workerConfig.getSecretsProviderConfiguratorConfig());
 
-        Optional<FunctionAuthProvider> functionAuthProvider = Optional.empty();
```
Why do we need to change this?

addisonj replied:
Because we now allow namespaces to be overridden by the ManifestCustomizer, we end up with an authprovider class being created per function, as the authProvider might be writing creds to different namespaces depending on the function. We do that by delaying the instantiation of the AuthProvider until we have the function details, so we just pass the classname into runtime providers.

jerrypeng commented on pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/kubernetes/KubernetesRuntimeFactory.java:
```diff
@@ -181,24 +186,23 @@ public void initialize(WorkerConfig workerConfig, AuthenticationConfig authentic
 
-        if (functionAuthProvider.isPresent()) {
-            if (!(functionAuthProvider.get() instanceof KubernetesFunctionAuthProvider)) {
+        this.authProviderClassName = functionAuthProviderClass;
```
I don't think we need change this

addisonj replied:
see comment here https://github.com/apache/pulsar/pull/5400/files#r344416501

jerrypeng commented on pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/kubernetes/KubernetesRuntimeFactory.java:
```diff
@@ -267,7 +281,8 @@ public void close() {
         KubernetesRuntime.doChecks(functionDetails);
         validateMinResourcesRequired(functionDetails);
-        secretsProviderConfigurator.doAdmissionChecks(appsClient, coreClient, jobNamespace, functionDetails);
+        KubernetesManifestCustomizer manifestCustomizer = buildManifestCustomizer(functionDetails);
```
Why do we need to add this here?

jerrypeng commented on pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/kubernetes/KubernetesRuntimeFactory.java:
```diff
@@ -361,8 +376,26 @@ void validateMinResourcesRequired(Function.FunctionDetails functionDetails) {
     @Override
-    public Optional<FunctionAuthProvider> getAuthProvider() {
-        return Optional.ofNullable(authProvider.orElse(null));
+    public Optional<FunctionAuthProvider> getAuthProvider(Function.FunctionDetails funcDetails) {
```
lets not change this

addisonj replied:
Yeah, so the problem is (or come up with an alternative) the issues is that we either:

1. The `KubernetesFunctionAuthProvider` won't know the right namespace to create the secret in. Since secrets are per namespace, that will break for any functions that override the namespace they want to run in. Or
2. We could call change the `KubernetesFunctionAuthProvider` such that it gets the namespace via some other means (instead of the initialize method, obviously, we don't want to have the initialize method called multiple times as that would be very strange for custom implementations). Looking at the API, I don't see any clean way to add the namespace into the `cacheAuthData` call of the FunctionAuthProvider. I spent a fair amount of time considering alternatives, but this seemed like the cleanest approach to simple delay the invocation of the AuthProvider and have it be scoped per function.

jerrypeng commented on pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/kubernetes/KubernetesRuntimeFactory.java:
```diff
@@ -104,7 +109,7 @@
     @ToString.Exclude
     @EqualsAndHashCode.Exclude
-    private Optional<KubernetesFunctionAuthProvider> authProvider;
+    private Optional<String> authProviderClassName;
```
Why do we need to change this?

addisonj replied:
See comment here: https://github.com/apache/pulsar/pull/5400/files#r344416501

jerrypeng commented on pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/BasicKubernetesManifestCustomizer.java:
```diff
@@ -0,0 +1,107 @@
+    private RuntimeOpts opts;
+
+    @Override
+    public void init(Function.FunctionDetails funcDetails, String staticConfig) {
```
staticConfig is not used?

addisonj replied:
Not in the implementation, but our use case for this is basically to have certain tenants to be mapped to specific namespaces and what not, so in that case, we could pass some configuration that would contain this mapping (rather than having to hardcode it in our implementation)

addisonj commented:
I considered such an approach, but wasn't sure if that was something that made sense for other providers. Will take a look at refactoring this to support that. However, I do believe the changes (or som other changes) to KubernetesFunctionAuthProvider are necessary such that the namespace in which the secret is created can be overridden by the customizer.

jerrypeng commented:
@addisonj also see if modiying "KubernetesFunctionAuthProvider/FunctionAuthProvider" interface to take in custom options can satisfy your use case.

addisonj commented:
Perhaps I am missing something, but I think  there is a mismatch here. The `FunctionRuntimeManager` can manage the creation of the `RuntimeCustomizer`, but it couldn't be eagerly in it's constructor or initialization methods. It would need to be during the `addAssignment` and `updateAssignment` methods as that is the first place where we get the function details.

In other words, the `FunctionRuntimeManager`'s lifecycle is such that it managed multiple functions and would need to have a factory method to create a `RuntimeCustomizer` for each function. This is essentially the problem with `FunctionAuthProvider` as it currently is created once for the whole runtime and instead needs to be per function.

Perhaps the `RuntimeCustomizer` could be responsible for the lifecycle of the `FunctionAuthProvider` as well, but their lifecycles would seem intertwined. 

Let me know if I am missing something.

addisonj commented on pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/auth/KubernetesFunctionAuthProvider.java:
```diff
@@ -30,7 +32,7 @@
 public interface KubernetesFunctionAuthProvider extends FunctionAuthProvider {
 
-    void initialize(CoreV1Api coreClient, String kubeNamespace, byte[] caBytes);
+    void initialize(CoreV1Api coreClient, byte[] caBytes, java.util.function.Function<Function.FunctionDetails, String> namespaceCustomizerFunc);
```
Now, instead of taking the namespace directly, it has a call to build the namespace from the `FunctionDetails`. This allows us to have it call the `KubernetesManifestCustomizer` via this function but doesn't require it to directly have knowledge of it.

jerrypeng replied:
Since caBytes and namespaceCustomizerFunc are essentially optional arguments, lets add setters for them to be with default being a no-op so that the interface can be cleaner.  Not all implementations will require caBytes and namespaceCustomizerFunc.

addisonj replied:
done

addisonj commented on pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/auth/FunctionAuthProvider.java:
```diff
@@ -39,29 +40,22 @@
-
-
-    Optional<FunctionAuthData> updateAuthData(String tenant, String namespace, String name, Optional<FunctionAuthData> existingFunctionAuthData, AuthenticationDataSource authenticationDataSource) throws Exception;
+    Optional<FunctionAuthData> cacheAuthData(Function.FunctionDetails funcDetails, AuthenticationDataSource authenticationDataSource) throws Exception;
```
The changes to these methods allow us the flexibility we need to have implementations of this interface, primarily the `KubernetesSecretTokenAuthProvider`, have all context it needs to make sure that secrets it generates match the kubenetes namespace the function actually gets deployed too. 

This also seems like a reasonable thing to do so that implementations of this interface can have more context about the function for custom implementations

addisonj commented on pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeFactory.java:
```diff
@@ -55,7 +56,11 @@ Runtime createContainer(
     default void doAdmissionChecks(Function.FunctionDetails functionDetails) { }
 
-    default Optional<FunctionAuthProvider> getAuthProvider() {
+    default Optional<? extends FunctionAuthProvider> getAuthProvider() {
```
This change allows implementations of this interface to provide a more specific FunctionAuthProvider, which reduces the need for casting

addisonj commented:
@jerrypeng I ended up re-working this a fair amount to be more in-line with some of your ideas. Primarily:

- created the `RuntimeCustomizer` class and had the `KubernetesManifestCustomizer` extend that class
- the `RuntimeCustomizer` now gets created in the `FunctionRuntimeManager` and is passed to all `RuntimeFactory` classes 
- reverted the changes to how the `FunctionAuthProvider` gets created to also be back in `FunctionRuntimeManager`

To make this work, it required a few changes to some of the interfaces, primarily:

- The `FunctionAuthProvider` now takes the `FunctionDetails`, this gives us all the data we need to effectively have the `FunctionAuthProvider` call the `KubernetesManifestCustomizer` so that we could get at the `customRuntimeOptions` which may change the namespace that secrets are created under. Additionally, this seems like a sane thing to do as now implementors of that interface get more data, which could be useful for different types of implementations
- Made some small tweaks to the `RuntimeFactory` such that the returned `FunctionAuthProvider` and `RuntimeCustomzer` can be sub-classes of those classes that are specialized by implementing interfaces
- Changed the `KubernetesManifestCustomizer` such that it only needs one instance of it as all the data needed to customize is passed via the methods (basically added `FunctionDetails` being passed to all calls)

jerrypeng commented on pulsar-functions/runtime/src/main/java/org/apache/pulsar/functions/runtime/RuntimeCustomizer.java:
```diff
@@ -0,0 +1,29 @@
+import org.apache.pulsar.functions.utils.Reflections;
+
+public interface RuntimeCustomizer {
+    void initialize(String staticConfig);
```
This "staticConfig" serves as a config for the a RuntimeCustomizer plugin.  Currently it's very vague what the purpose of it should be. Let's just call it runtimeCustomizerConfig or just config and has it be a final Map<String, Object> config.

jerrypeng commented:
@addisonj thanks for taking the time to rework some of this PR!  It looks really good.  I left some comments.

addisonj commented:
@jerrypeng changes made, let me know if you have anything else, glad it shaped up well :)

addisonj commented:
rerun integration tests

addisonj commented:
rerun integration tests

addisonj commented:
@sijie if you have a second to merge this, that would be awesome :)

sijie commented:
@addisonj great contribution!
